backend/YouTubeV2.Application/Services/SearchService.cs

```diff
@@ -0,0 +1,98 @@
+﻿using Microsoft.AspNetCore.Identity;
+using YouTubeV2.Application.DTO.SearchDTOS;
+using YouTubeV2.Application.DTO.UserDTOS;
+using YouTubeV2.Application.Enums;
+using YouTubeV2.Application.Exceptions;
+using YouTubeV2.Application.Model;
+
+namespace YouTubeV2.Application.Services
+{
+    public class SearchService : ISearchService
+    {
+        private readonly UserManager<User> _userManager;
+        private readonly IUserService _userService;
+        private readonly ISubscriptionService _subscriptionService;
+
+        public SearchService(UserManager<User> userManager, IUserService userService, ISubscriptionService subscriptionService) 
+        {
+            _userManager = userManager;
+            _userService = userService;
+            _subscriptionService = subscriptionService;
+        }
+
+        public async Task<SearchResultsDto> SearchAsync(string query, SortingDirections sortingDirection,
+            SortingTypes sortingType, DateTimeOffset? dateBegin, DateTimeOffset? dateEnd, CancellationToken cancellationToken)
+        {
+            var users = await SearchForUsersAsync(query, sortingDirection, sortingType, dateBegin, dateEnd, cancellationToken);
+
+            return new SearchResultsDto(users);
+        }
+
+        private async Task<IReadOnlyList<UserDto>> SearchForUsersAsync(string query, SortingDirections sortingDirection,
+            SortingTypes sortingType, DateTimeOffset? dateBegin, DateTimeOffset? dateEnd, CancellationToken cancellationToken)
+        {
+            var searchableUsers = await _userManager.GetUsersInRoleAsync(Role.Creator);
+            var matchingUsers = searchableUsers.Select(user => user).Where(user => user.UserName!
+                .Contains(query, StringComparison.InvariantCultureIgnoreCase)).ToList();
```

Comment:
I think we shouldnt call ToList here. It is because how ef works. The collection doesnt get loaded to memory till u materialize it with function like ToList. Since we are to clip the users based on date then we should do it on IQueryable and then materialize it. idk about sorting though. I guess if we can sort IQueryable we should do it. And use ToListAsync

Reply:
i changed most of serach logic, tell me if it is ok now